kernel/src/process_checking.rs

```diff
@@ -0,0 +1,355 @@
+                    return true;
+                } else {
+                    let data_a = a.data();
+                    let data_b = b.data();
+                    for (p1, p2) in data_a.iter().zip(data_b.iter()) {
+                        if p1 != p2 {
+                            return true;
+                        }
+                    }
+                }
+                false
+            })
+        })
+    }
+}
+
+impl ClientData<32_usize> for AppCheckerSha256 {
+    fn add_mut_data_done(
+        &self,
+        _result: Result<(), ErrorCode>,
+        _data: LeasableMutableBuffer<'static, u8>,
+    ) {
+    }
+
+    fn add_data_done(&self, result: Result<(), ErrorCode>, data: LeasableBuffer<'static, u8>) {
+        match result {
+            Err(e) => panic!("Internal error during application binary checking. SHA256 engine threw error in adding data: {:?}", e),
+            Ok(()) => {
+                self.binary.set(data.take());
+                let hash: &'static mut [u8; 32_usize] = self.hash.take().unwrap();
+                match self.hasher.verify(hash) {
+                    Err((e, _)) => panic!("Failed invoke hash verification in process credential checking: {:?}", e),
+                    Ok(()) => {},
+                }
+            }
+        }
+    }
+}
+
+impl<'a> ClientVerify<32_usize> for AppCheckerSha256 {
+    fn verification_done(
+        &self,
+        result: Result<bool, ErrorCode>,
+        compare: &'static mut [u8; 32_usize],
+    ) {
+        self.hash.replace(compare);
+        match result {
+            Ok(true) => {
+                self.client.map(|c| {
+                    c.check_done(
+                        Ok(CheckResult::Accept),
+                        self.credentials.take().unwrap(),
+                        self.binary.take().unwrap(),
+                    );
+                });
+            }
+            Ok(false) => {
+                self.client.map(|c| {
+                    c.check_done(
+                        Ok(CheckResult::Reject),
+                        self.credentials.take().unwrap(),
+                        self.binary.take().unwrap(),
+                    );
+                });
+            }
+            Err(e) => {
+                panic!("Error {:?} in processing application credentials.", e);
+            }
+        }
+    }
+}
+
+impl<'a> ClientHash<32_usize> for AppCheckerSha256 {
+    fn hash_done(&self, _result: Result<(), ErrorCode>, _digest: &'static mut [u8; 32_usize]) {}
+}
+
+impl Compress for AppCheckerSha256 {
+    fn to_short_id(&self, credentials: &TbfFooterV2Credentials) -> Option<ShortID> {
+        let id: u32 = (credentials.data()[0] as u32) << 24
+            | (credentials.data()[1] as u32) << 16
+            | (credentials.data()[2] as u32) << 8
+            | (credentials.data()[3] as u32);
+        match core::num::NonZeroU32::new(id) {
+            Some(nzid) => Some(ShortID { id: nzid }),
+            None => None,
```

Comment:
Not sure I understand this part correctly. This looks to me like there is a 1 in `u32::MAX` chance to end up with a verified but compressed `AppId` which is unrepresentable, is that correct?

It might be cleaner to instead take only 31 bits of the resulting hash, and set one bit to constant `1`, such that every hash has a corresponding valid (and possibly colliding) compressed `ShortID` representation.

Reply:
Good point. I will fix this. This was checker was intended to just be a simple demonstration, but it would be better to not have this issue.